governance/validatorSet.go

```diff
@@ -0,0 +1,185 @@
+	cfg := txclient.NewTxDefaultConfig(deployContractGasLimit)
+
+	txOps, err := txclient.GenerateTxOpts(ctx, client, txAuth, cfg)
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	addr, _, _, err := bindings.DeployValidatorSet(txOps, client)
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	return addr, nil
+}
+
+func (v ValidatorSet) AddValidator(txAuth authy.Auth, pubKey string, address common.Address) (error) {
+	client, err := ethclient.Dial(v.gethIpc)
+	if err != nil {
+		return err
+	}
+
+	defer client.Close()
+
+	contractInstance, err := bindings.NewValidatorSet(v.contractAddress, client)
+	if err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+	cfg := txclient.NewTxDefaultConfig(deployContractGasLimit)
+	txOps, err := txclient.GenerateTxOpts(ctx, client, txAuth, cfg)
+	if err != nil {
+		return err
+	}
+
+	tx, err := contractInstance.AddValidator(txOps, pubKey, address)
+	if err != nil {
+		return err
+	}
+	
+	_, err = txclient.FetchReceipt(client, tx, cfg)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (v ValidatorSet) RemoveValidator(txAuth authy.Auth, pubKey string, address common.Address) (error) {
+	client, err := ethclient.Dial(v.gethIpc)
+	if err != nil {
+		return err
+	}
+
+	defer client.Close()
+
+	contractInstance, err := bindings.NewValidatorSet(v.contractAddress, client)
+	if err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+	cfg := txclient.NewTxDefaultConfig(deployContractGasLimit)
+	txOps, err := txclient.GenerateTxOpts(ctx, client, txAuth, cfg)
+	if err != nil {
+		return err
+	}
+
+	tx, err := contractInstance.RemoveValidator(txOps, pubKey, address)
+	if err != nil {
+		return err
+	}
+	
+	_, err = txclient.FetchReceipt(client, tx, cfg)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (v ValidatorSet) FetchPubKeySet(address common.Address) ([]string, error) {
+	client, err := ethclient.Dial(v.gethIpc)
+	if err != nil {
+		return nil, err
+	}
+
+	defer client.Close()
+
+	contractInstance, err := bindings.NewValidatorSet(v.contractAddress, client)
+	if err != nil {
+		return nil, err
+	}
+
+	validatorSetSizeBN, err := contractInstance.ValidatorSetSize(&bind.CallOpts{})
+	if err != nil {
+		return nil, err
+	}
+	validatorSetSize :=int(validatorSetSizeBN.Int64())
+	var validatorSet []string
+	for i := 0; i < validatorSetSize; i++ {
+		validatorPubKey, err := contractInstance.ValidatorPubKey(&bind.CallOpts{}, big.NewInt(int64(i)))
+		if err != nil {
+			return nil, err
+		}
+		validatorSet = append(validatorSet, validatorPubKey)
+	}
+
+	return validatorSet, nil
+}
+
+func (v ValidatorSet) ValidatorAddress(pubKey string) (common.Address, error) {
+	if v.contractAddress.String() == common.HexToAddress("0x0").String() {
+		return common.Address{}, nil
+	}
+
+	client, err := ethclient.Dial(v.gethIpc)
+	if err != nil {
+		return common.Address{}, err
+	}
+	defer client.Close()
+
+	contractInstance, err := bindings.NewValidatorSetCaller(v.contractAddress, client)
```

Comment:
[critical] Correct me if I am wrong but I think this is not going to work. 

This contract address is coming from a hard-coded config. Which means every time a validator boots up it connects to this contract. This is going to break a lot if a validator didn't upgrade it's lightchain version. BUT you smartly introduced the concept of versioning in ValidatorSet SC using a "nextVersion" attribute to solve this issue.

I would, we should unify all the "NewValidatorSet(address)" to a function that would iterate via linked-list style through all the SC instances until nextVersion != 0x0. In that time we will know we are using the latest version of the SC from the software perspective. And this would work even if some validator wouldn't upgrade it's latest version of Lightchain as on a bootup they would start in version "1" on address 0x111111 with nextVersion set to "0x222222" then it would open the version 0x22222 check for nextVersion etc.

WDYT?

Reply:
Let's have a chat offline. 

Yes, the concept of version should sort this problem out, but I didn't intend to work around that topic on this PR.

Reply:
As it is explained on Tendermint specs the pubKey should be storable in 20 bytes https://github.com/tendermint/tendermint/blob/master/docs/spec/blockchain/encoding.md. 

I got to sort the issue and simplify the smart contract by the usage of a byte20 type

Reply:
ok great